File: src/data/tutorials/079-daily-temperatures.js
/**
 * Tutorial content for #079 — Daily Temperatures
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an array of daily temperatures, return an array where each element tells you how many days you have to wait until a warmer temperature. If no future day is warmer, put 0 for that day.`,
    example: `temps = [73,74,75,71,69,72,76,73]\nDay 0 (73°): next warmer is day 1 (74°) → 1 day wait\nDay 1 (74°): next warmer is day 2 (75°) → 1 day wait\nDay 2 (75°): next warmer is day 6 (76°) → 4 days wait\nDay 3 (71°): next warmer is day 5 (72°) → 2 days wait\n✅ Answer: [1,1,4,2,1,1,0,0]`,
    keyInsight: `You need the "next greater element" to the right for each position. A monotonic stack (decreasing) lets you solve this in one pass: stack stores indices of days whose warmer future day hasn't been found yet.`,
  },

  approaches: {
    'Monotonic Stack (Optimal)': {
      intuition: `Keep a stack of indices of days that are still "waiting" for a warmer day. The stack stays in decreasing order of temperature (monotonically decreasing).\n\nFor each new day, if its temperature is warmer than the top of the stack, we've found the answer for that waiting day — pop it, compute the gap, record it. Keep popping while the new day is warmer. Then push the current index.`,
      steps: [
        `Initialize result array of zeros, same length as temperatures.`,
        `Create an empty stack (stores indices).`,
        `For each index i:`,
        `  While stack is not empty AND temps[i] > temps[stack.top]:`,
        `    idx = stack.pop()`,
        `    result[idx] = i - idx`,
        `  Push i onto the stack.`,
        `Return result (any index still in the stack stays 0).`,
      ],
      example: `temps = [73,74,75,71,69,72,76,73]\nresult = [0,0,0,0,0,0,0,0]\n\ni=0(73): stack empty → push 0.  stack:[0]\ni=1(74): 74>73 → pop 0, res[0]=1-0=1. push 1. stack:[1]\ni=2(75): 75>74 → pop 1, res[1]=2-1=1. push 2. stack:[2]\ni=3(71): 71<75 → push 3.  stack:[2,3]\ni=4(69): 69<71 → push 4.  stack:[2,3,4]\ni=5(72): 72>69 → pop 4, res[4]=5-4=1\n         72>71 → pop 3, res[3]=5-3=2\n         72<75 → stop. push 5. stack:[2,5]\ni=6(76): 76>72 → pop 5, res[5]=6-5=1\n         76>75 → pop 2, res[2]=6-2=4\n         stack empty → push 6. stack:[6]\ni=7(73): 73<76 → push 7. stack:[6,7]\nResult: [1,1,4,2,1,1,0,0] ✅`,
      keyInsight: `O(n) time — each index is pushed and popped at most once. O(n) space for the stack. This is the canonical "next greater element" pattern and appears frequently in interview problems.`,
    },

    'Brute Force': {
      intuition: `For each day i, scan every future day j (j > i) until you find a temperature warmer than temps[i]. Record the gap j - i. Simple to understand but slow for large inputs.`,
      steps: [
        `Initialize result array of zeros.`,
        `For each index i from 0 to n-2:`,
        `  For each index j from i+1 to n-1:`,
        `    If temps[j] > temps[i]: result[i] = j - i; break.`,
        `Return result.`,
      ],
      example: `temps = [73,74,75,71]\n\ni=0(73): j=1(74)>73 → res[0]=1. stop.\ni=1(74): j=2(75)>74 → res[1]=1. stop.\ni=2(75): j=3(71)<75 → no warmer found → res[2]=0.\ni=3(71): no j → res[3]=0.\nResult: [1,1,0,0] ✅`,
      keyInsight: `O(n²) time, O(1) extra space. Fine for n ≤ 100 but too slow for n = 100,000. Useful as a correctness reference to verify faster approaches.`,
    },

    'Backward Iteration with Jump': {
      intuition: `Iterate from the second-to-last day backward to day 0. For each day i, use the already-computed result array to "jump" forward: if temps[j] ≤ temps[i], jump by result[j] steps to find the next candidate, repeating until you find a warmer day or fall off the end.`,
      steps: [
        `Initialize result array of zeros.`,
        `Iterate i from n-2 down to 0:`,
        `  j = i + 1.`,
        `  While j < n AND temps[j] <= temps[i]:`,
        `    If result[j] == 0: break (no warmer day exists beyond j).`,
        `    j += result[j]  ← jump forward using precomputed answer.`,
        `  If j < n AND temps[j] > temps[i]: result[i] = j - i.`,
        `Return result.`,
      ],
      example: `temps = [73,74,75,71,69,72,76,73]\n\ni=6(76): j=7(73) ≤ 76, result[7]=0 → break. res[6]=0.\ni=5(72): j=6(76) > 72 → res[5]=1.\ni=4(69): j=5(72) > 69 → res[4]=1.\ni=3(71): j=4(69) ≤ 71, jump j+=1→5(72) > 71 → res[3]=2.\ni=2(75): j=3, jump→5, jump→6(76)>75 → res[2]=4.\ni=1(74): j=2, jump→6(76)>74 → res[1]=1.\ni=0(73): j=1, jump→2, jump→6(76)>73 → res[0]=1.\nResult: [1,1,4,2,1,1,0,0] ✅`,
      keyInsight: `O(n) amortized time, O(1) extra space (uses only the output array). The jump avoids re-scanning already-resolved days. A clever space-saving alternative to the monotonic stack.`,
    },

    'Array as Stack': {
      intuition: `Exactly the monotonic stack approach, but uses a plain integer array with a manual top pointer instead of a Stack/Deque object. This is the same algorithm — just a micro-optimization to avoid object allocation overhead in languages like Java.`,
      steps: [
        `int[] stack = new int[n]; int top = -1.`,
        `int[] result = new int[n].`,
        `For each i from 0 to n-1:`,
        `  While top >= 0 AND temps[i] > temps[stack[top]]:`,
        `    idx = stack[top--].`,
        `    result[idx] = i - idx.`,
        `  stack[++top] = i.`,
        `Return result.`,
      ],
      example: `temps = [73,74,75,71]\nstack=[], top=-1\n\ni=0: push 0. stack=[0]\ni=1(74>73): pop 0 res[0]=1, push 1. stack=[1]\ni=2(75>74): pop 1 res[1]=1, push 2. stack=[2]\ni=3(71<75): push 3. stack=[2,3]\nResult: [1,1,0,0] ✅`,
      keyInsight: `O(n) time, O(n) space. Identical to the Monotonic Stack approach in behavior. Using an array manually is common in competitive programming for speed; in an interview, the Deque version is more readable.`,
    },

    'Next Greater Element Pattern': {
      intuition: `Recognize this as the classic "Next Greater Element" problem. Some implementations solve it by processing the stack in a slightly different order — pushing (temperature, index) pairs — or by doing two passes. The core idea is unchanged: a monotonic decreasing stack resolves waiting indices as soon as a larger element arrives.`,
      steps: [
        `Stack stores (temperature, index) pairs.`,
        `For each (i, temp) in enumerated temperatures:`,
        `  While stack not empty AND temp > stack.top.temperature:`,
        `    (t, idx) = stack.pop()`,
        `    result[idx] = i - idx`,
        `  Push (temp, i).`,
        `Return result.`,
      ],
      example: `temps = [73,74,75,71]\n\ni=0,T=73: push (73,0). stack:[(73,0)]\ni=1,T=74: 74>73 → pop(73,0) res[0]=1-0=1. push (74,1).\ni=2,T=75: 75>74 → pop(74,1) res[1]=2-1=1. push (75,2).\ni=3,T=71: 71<75 → push (71,3).\nStack remaining: (75,2),(71,3) → res stays 0.\nResult: [1,1,0,0] ✅`,
      keyInsight: `O(n) time, O(n) space. Storing pairs is slightly more memory per entry but makes the code more explicit. Functionally equivalent to storing index-only — the temperature can always be recovered via temps[idx].`,
    },
  },
}

File: public/cpp/079-daily-temperatures.cpp
/*
 * Problem: Daily Temperatures (LeetCode 739)
 * Link: https://leetcode.com/problems/daily-temperatures/
 * Difficulty: Medium
 * Category: Stacks and Queues
 * 
 * Description:
 * Given an array of integers temperatures represents the daily temperatures,
 * return an array answer such that answer[i] is the number of days you have to
 * wait after the ith day to get a warmer temperature. If there is no future day
 * for which this is possible, keep answer[i] == 0 instead.
 * 
 * Example 1:
 * Input: temperatures = [73,74,75,71,69,72,76,73]
 * Output: [1,1,4,2,1,1,0,0]
 * 
 * Example 2:
 * Input: temperatures = [30,40,50,60]
 * Output: [1,1,1,0]
 * 
 * Example 3:
 * Input: temperatures = [30,60,90]
 * Output: [1,1,0]
 * 
 * Constraints:
 * - 1 <= temperatures.length <= 10^5
 * - 30 <= temperatures[i] <= 100
 */

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

/*
 * APPROACH 1: MONOTONIC STACK (OPTIMAL)
 * 
 * Intuition:
 * - Use stack to track indices of temperatures waiting for warmer day
 * - Stack maintains decreasing temperature order
 * - When we find warmer temperature, pop all smaller temperatures from stack
 * - Calculate days difference for each popped index
 * 
 * Algorithm:
 * 1. Iterate through temperatures
 * 2. While stack not empty and current temp > stack top temp:
 *    - Pop index from stack
 *    - Calculate days = current index - popped index
 *    - Store in result
 * 3. Push current index to stack
 * 4. Return result
 * 
 * Time Complexity: O(n) - each element pushed and popped once
 * Space Complexity: O(n) - stack size
 */
class Solution1 {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> result(n, 0);
        stack<int> stk; // Store indices
        
        for (int i = 0; i < n; i++) {
            // Pop all temperatures smaller than current
            while (!stk.empty() && temperatures[i] > temperatures[stk.top()]) {
                int prevIndex = stk.top();
                stk.pop();
                result[prevIndex] = i - prevIndex;
            }
            
            stk.push(i);
        }
        
        return result;
    }
};

/*
 * APPROACH 2: BRUTE FORCE (FOR COMPARISON)
 * 
 * Intuition:
 * - For each day, search forward for warmer temperature
 * - Simple but inefficient
 * - Good for understanding the problem
 * 
 * Algorithm:
 * 1. For each index i:
 *    - Search from i+1 to end
 *    - Find first temperature > temperatures[i]
 *    - Calculate days difference
 * 2. Return result
 * 
 * Time Complexity: O(n^2) - nested loops
 * Space Complexity: O(1) - excluding output
 */
class Solution2 {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> result(n, 0);
        
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (temperatures[j] > temperatures[i]) {
                    result[i] = j - i;
                    break;
                }
            }
        }
        
        return result;
    }
};

/*
 * APPROACH 3: BACKWARD ITERATION WITH OPTIMIZATION
 * 
 * Intuition:
 * - Iterate from right to left
 * - Use result array to jump to next warmer day
 * - If result[j] == 0, no warmer day exists
 * - Otherwise, jump by result[j] days
 * 
 * Algorithm:
 * 1. Start from second last element
 * 2. For each position i:
 *    - Start checking from i+1
 *    - If temp[j] > temp[i], found answer
 *    - If temp[j] <= temp[i] and result[j] == 0, no answer
 *    - Otherwise, jump to j + result[j]
 * 3. Return result
 * 
 * Time Complexity: O(n) - amortized, each element visited constant times
 * Space Complexity: O(1) - excluding output
 */
class Solution3 {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> result(n, 0);
        
        for (int i = n - 2; i >= 0; i--) {
            int j = i + 1;
            
            while (j < n) {
                if (temperatures[j] > temperatures[i]) {
                    result[i] = j - i;
                    break;
                }
                
                if (result[j] == 0) {
                    break; // No warmer day exists
                }
                
                j += result[j]; // Jump to next potential warmer day
            }
        }
        
        return result;
    }
};

/*
 * APPROACH 4: ARRAY AS STACK
 * 
 * Intuition:
 * - Use array instead of stack for better cache locality
 * - Maintain stack pointer manually
 * - Same logic as monotonic stack
 * 
 * Algorithm:
 * 1. Use array to simulate stack
 * 2. Track stack size with pointer
 * 3. Same monotonic stack logic
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
class Solution4 {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> result(n, 0);
        vector<int> stack(n);
        int top = -1;
        
        for (int i = 0; i < n; i++) {
            while (top >= 0 && temperatures[i] > temperatures[stack[top]]) {
                int prevIndex = stack[top--];
                result[prevIndex] = i - prevIndex;
            }
            
            stack[++top] = i;
        }
        
        return result;
    }
};

/*
 * APPROACH 5: NEXT GREATER ELEMENT PATTERN
 * 
 * Intuition:
 * - This is a variant of "Next Greater Element" problem
 * - Use same monotonic stack pattern
 * - Store both index and temperature in stack
 * 
 * Algorithm:
 * 1. Use stack of pairs (index, temperature)
 * 2. For each element, pop smaller temperatures
 * 3. Calculate days for each popped element
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
class Solution5 {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> result(n, 0);
        stack<pair<int, int>> stk; // (index, temperature)
        
        for (int i = 0; i < n; i++) {
            while (!stk.empty() && temperatures[i] > stk.top().second) {
                auto [prevIndex, prevTemp] = stk.top();
                stk.pop();
                result[prevIndex] = i - prevIndex;
            }
            
            stk.push({i, temperatures[i]});
        }
        
        return result;
    }
};

// Test function
void test(vector<int> temperatures, int approach) {
    vector<int> result;
    
    cout << "Input: [";
    for (int i = 0; i < temperatures.size(); i++) {
        cout << temperatures[i];
        if (i < temperatures.size() - 1) cout << ",";
    }
    cout << "]\n";
    
    switch(approach) {
        case 1: {
            Solution1 sol;
            result = sol.dailyTemperatures(temperatures);
            cout << "Approach 1 (Monotonic Stack): ";
            break;
        }
        case 2: {
            Solution2 sol;
            result = sol.dailyTemperatures(temperatures);
            cout << "Approach 2 (Brute Force): ";
            break;
        }
        case 3: {
            Solution3 sol;
            result = sol.dailyTemperatures(temperatures);
            cout << "Approach 3 (Backward Iteration): ";
            break;
        }
        case 4: {
            Solution4 sol;
            result = sol.dailyTemperatures(temperatures);
            cout << "Approach 4 (Array as Stack): ";
            break;
        }
        case 5: {
            Solution5 sol;
            result = sol.dailyTemperatures(temperatures);
            cout << "Approach 5 (Next Greater Pattern): ";
            break;
        }
    }
    
    cout << "[";
    for (int i = 0; i < result.size(); i++) {
        cout << result[i];
        if (i < result.size() - 1) cout << ",";
    }
    cout << "]\n\n";
}

int main() {
    // Test Case 1: Standard case
    cout << "Test Case 1: Standard case\n";
    vector<int> test1 = {73, 74, 75, 71, 69, 72, 76, 73};
    for (int i = 1; i <= 5; i++) {
        test(test1, i);
    }
    
    // Test Case 2: Increasing temperatures
    cout << "Test Case 2: Increasing\n";
    vector<int> test2 = {30, 40, 50, 60};
    for (int i = 1; i <= 5; i++) {
        test(test2, i);
    }
    
    // Test Case 3: Decreasing temperatures
    cout << "Test Case 3: Decreasing\n";
    vector<int> test3 = {90, 80, 70, 60};
    for (int i = 1; i <= 5; i++) {
        test(test3, i);
    }
    
    // Test Case 4: Single element
    cout << "Test Case 4: Single element\n";
    vector<int> test4 = {30};
    for (int i = 1; i <= 5; i++) {
        test(test4, i);
    }
    
    // Test Case 5: All same
    cout << "Test Case 5: All same\n";
    vector<int> test5 = {30, 30, 30};
    for (int i = 1; i <= 5; i++) {
        test(test5, i);
    }
    
    return 0;
}

/*
 * COMPLEXITY ANALYSIS SUMMARY:
 * 
 * Approach 1 (Monotonic Stack):
 * - Time: O(n) - each element pushed/popped once
 * - Space: O(n) - stack size
 * - Best for: Optimal solution, most common
 * 
 * Approach 2 (Brute Force):
 * - Time: O(n^2) - nested loops
 * - Space: O(1)
 * - Best for: Understanding problem (not optimal)
 * 
 * Approach 3 (Backward Iteration):
 * - Time: O(n) - amortized
 * - Space: O(1) - excluding output
 * - Best for: Space-efficient solution
 * 
 * Approach 4 (Array as Stack):
 * - Time: O(n)
 * - Space: O(n)
 * - Best for: Better cache locality
 * 
 * Approach 5 (Next Greater Pattern):
 * - Time: O(n)
 * - Space: O(n)
 * - Best for: Showing pattern recognition
 * 
 * INTERVIEW TIPS:
 * 1. Recognize as "Next Greater Element" variant
 * 2. Start with Approach 1 (monotonic stack)
 * 3. Explain why stack maintains decreasing order
 * 4. Mention Approach 3 for space optimization
 * 5. Draw example to visualize stack operations
 * 
 * COMMON MISTAKES:
 * 1. Using max heap instead of stack (wrong data structure)
 * 2. Not initializing result array with zeros
 * 3. Storing temperatures instead of indices in stack
 * 4. Off-by-one errors in day calculation
 * 5. Not handling case where no warmer day exists
 * 
 * FOLLOW-UP QUESTIONS:
 * 1. What if we want previous warmer day? (Iterate left to right)
 * 2. Can you do it in O(1) space? (Yes, Approach 3)
 * 3. What if temperatures can be negative? (Same algorithm)
 * 4. How to find next k warmer days? (More complex, use heap)
 * 5. What if we want exact temperature difference? (Store temps)
 * 
 * RELATED PROBLEMS:
 * - Next Greater Element I
 * - Next Greater Element II (circular)
 * - Online Stock Span
 * - Largest Rectangle in Histogram
 * - Trapping Rain Water
 * 
 * KEY INSIGHTS:
 * 1. Monotonic stack is key data structure
 * 2. Stack maintains decreasing temperature order
 * 3. Each element processed exactly once
 * 4. Backward iteration can save space
 * 5. This is a classic stack application
 * 
 * MONOTONIC STACK PATTERN:
 * - Used for "next greater/smaller element" problems
 * - Maintains increasing or decreasing order
 * - O(n) time complexity
 * - Each element pushed and popped at most once
 * - Very common in interviews
 */

// Made with Bob
